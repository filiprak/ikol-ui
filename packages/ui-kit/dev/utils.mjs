import { existsSync } from 'fs';
import { readFile, writeFile, mkdir } from 'fs/promises';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

export async function useTemplateFile(filename, outputFilepath, variables) {
  const data = await readFile(path.resolve(__dirname, 'templates', filename), 'utf8');
  return useTemplate(data, outputFilepath, variables);
}

export async function useTemplate(templateStr, outputFilepath, variables) {
  const updatedTemplate = templateStr.replace(/\[\[([a-zA-Z0-9_]+)\]\]/g, (match, key) => {
    return variables[key] || match;
  });

  if (!existsSync(path.dirname(outputFilepath))) {
    await mkdir(path.dirname(outputFilepath));
  }
  await writeFile(outputFilepath, updatedTemplate);
}

export const kebabize = str => {
  return str.split('').map((letter, idx) => {
      return letter.toUpperCase() === letter
          ? `${idx !== 0 ? '-' : ''}${letter.toLowerCase()}`
          : letter;
  }).join('');
}

export const ROOT_DIR = path.resolve(__dirname, '..');
