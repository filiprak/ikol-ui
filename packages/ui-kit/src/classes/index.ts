export type { IkTime } from './dates/IkTime';
export type { IkLocaleSettings } from './locale/settings';

export { time } from './dates';
