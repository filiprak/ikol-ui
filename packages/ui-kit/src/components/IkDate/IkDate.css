.ik-date {
    display: flex;
    flex-direction: row;
}

.ik-date__day-input {
    flex: 1 1 auto;
    min-width: 110px;
    max-width: 130px;
}

.ik-date__day-input:not(:last-child) {
    margin-right: 7px;
}

.ik-date--full-width .ik-date__day-input {
    max-width: 100%;
}

.ik-date__hour-input {
    flex: 0 1 auto;
    min-width: 50px;
    max-width: 80px;
}

.ik-date__calendar {
    padding: 13px;
    font-weight: 600;
}

.ik-date__row {
    display: flex;
    justify-content: center;
    align-items: center;
}

.ik-date__cell {
    display: inline-flex;
    justify-content: center;
    align-items: center;
    width: 38px;
    height: 36px;
    text-align: center;
}

.ik-date__cell--weekday {
    height: 24px;
    font-size: 11px;
    color: var(--on-surface-color-4);
}

.ik-date__cell--offset {
    color: var(--on-surface-color-3);
}

.ik-date__cell--active > div {
    background-color: var(--primary-color);
    color: var(--on-primary-color);
}

.ik-date__cell--today:not(.ik-date__cell--active) > div {
    color: var(--accent-color);
}

.ik-date__cell > div {
    display: inline-flex;
    justify-content: center;
    align-items: center;
    text-align: center;
    height: 32px;
    width: 32px;
    border-radius: 20px;
    cursor: pointer;
}

.ik-date__cell:not(.ik-date__cell--active) > div:hover {
    background-color: var(--primary-color-o1);
}

.ik-date__cell--disabled {
    pointer-events: none;
    opacity: 0.2;
}

.ik-date__nav {
    display: flex;
    padding-bottom: 7px;
}

.ik-date__nav-month {
    display: flex;
    text-align: center;
    justify-content: center;
    align-items: center;
    flex: 1 0 auto;
    text-transform: capitalize;
    font-size: 14px;
}

.ik-date__nav-btn {
    flex: 0 0 auto;
}

.ik-date__select {
    min-width: 120px;
    height: 200px;
    font-weight: 600;
    overflow: auto;
}

.ik-date__select-item {
    display: flex;
    align-items: center;
    cursor: pointer;
    padding: 8px 12px;
    height: 40px;
}

.ik-date__select-item:hover {
    background-color: var(--primary-color-o1);
    color: var(--primary-color);
}