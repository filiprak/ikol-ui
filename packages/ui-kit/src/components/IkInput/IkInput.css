.ik-input {
    position: relative;
}

.ik-input__append,
.ik-input__prepend {
    display: flex;
    align-items: center;
    justify-content: center;
    position: absolute;
    top: 0;
    bottom: 0;
    line-height: 24px;
    width: 32px;
    padding: 4px 8px;
    color: var(--on-background-color-3);
}

.ik-input__prepend {
    left: 0;
}

.ik-input__append {
    right: 0;
}

.ik-input--prepend > .ik-input__native {
    padding-left: 32px !important;
}

.ik-input--append > .ik-input__native {
    padding-right: 32px !important;
}

.ik-input:focus-within > .ik-input__prepend,
.ik-input:focus-within > .ik-input__append {
    color: var(--primary-color);
}

.ik-input--disabled > .ik-input__prepend,
.ik-input--disabled > .ik-input__append {
    opacity: 0.65;
    pointer-events: none;
}
