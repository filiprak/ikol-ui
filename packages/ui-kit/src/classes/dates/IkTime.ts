import type {
    Moment,
    unitOfTime,
    MomentInputObject,
} from 'moment-timezone';
import moment, {
    isMoment,
    tz,
} from 'moment-timezone';
import type { IkTimePreset } from '.';
import { time } from '.';
import { DEFAULT_FORMAT } from './formats';
import { isNullish, isString } from '@ui/utils/helpers';

export const DEFAULT_PARSE_FORMAT = 'YYYY-MM-DDTHH:mm:ssZ';

export type IkTimeInput = string | number | MomentInputObject | Date | Moment;

export class IkTime {
    private _preset: IkTimePreset = time;
    private _d: Moment;

    constructor(input?: IkTimeInput, format?: string, preset?: IkTimePreset) {
        if (preset) {
            this._preset = preset;
        }
        if (isNullish(input)) {
            input = new Date();
        }
        if (isMoment(input)) {
            this._d = input;
        } else {
            const source_tz = this.getTargetTimezone();

            if (input instanceof Date) {
                this._d = tz(new Date(), source_tz);
            } else if (isString(input)) {
                format = this.getLocalizedFormat(format || DEFAULT_PARSE_FORMAT)?.trim();
                this._d = tz(String(input || ''), format || DEFAULT_PARSE_FORMAT, source_tz);
            } else {
                this._d = tz(input, source_tz);
            }
        }
    }

    static utc(input?: IkTimeInput, format?: string, preset?: IkTimePreset) {
        if (isNullish(input)) {
            input = new Date();
        }
        if (isMoment(input)) {
            return new IkTime(input, undefined, preset);
        } else {
            if (input instanceof Date) {
                return new IkTime(moment.utc(), undefined, preset);
            } else if (isString(input)) {
                return new IkTime(moment.utc(String(input || ''), format || DEFAULT_PARSE_FORMAT), undefined, preset);
            } else {
                return new IkTime(moment.utc(input), undefined, preset);
            }
        }
    }

    format(format?: string) {
        format = String(format || DEFAULT_FORMAT).trim();

        if (this.isValid()) {
            return this
                .getDto()
                .clone()
                .tz(this.getTargetTimezone())
                .format(this.getLocalizedFormat(format));
        } else {
            return moment.invalid().format();
        }
    }

    formatUtc(format?: string) {
        format = String(format || DEFAULT_FORMAT).trim();

        if (this.isValid()) {
            return this
                .getDto()
                .tz('UTC')
                .format(this.getLocalizedFormat(format));
        } else {
            return moment.invalid().format();
        }
    }

    clone() {
        return new IkTime(this._d.clone(), undefined, this._preset);
    }

    isValid() {
        return this._d.isValid();
    }

    add(amount: number, unit: unitOfTime.DurationConstructor) {
        this._d = this._d.add(amount, unit);
        return this;
    }

    subtract(amount: number, unit: unitOfTime.DurationConstructor) {
        this._d = this._d.subtract(amount, unit);
        return this;
    }

    startOf(unit: unitOfTime.StartOf) {
        this._d = this._d.startOf(unit);
        return this;
    }

    endOf(unit: unitOfTime.StartOf) {
        this._d = this._d.endOf(unit);
        return this;
    }

    isBefore(input: IkTime, granularity?: unitOfTime.StartOf) {
        return this._d.isBefore(input.getDto(), granularity);
    }

    isAfter(input: IkTime, granularity?: unitOfTime.StartOf) {
        return this._d.isAfter(input.getDto(), granularity);
    }

    isSameOrAfter(input: IkTime, granularity?: unitOfTime.StartOf) {
        return this._d.isSameOrAfter(input.getDto(), granularity);
    }

    isSame(input: IkTime, granularity?: unitOfTime.StartOf) {
        return this._d.isSame(input.getDto(), granularity);
    }

    isSameOrBefore(input: IkTime, granularity?: unitOfTime.StartOf) {
        return this._d.isSameOrBefore(input.getDto(), granularity);
    }

    isBetween(from: IkTime, to: IkTime, granularity?: unitOfTime.StartOf, inclusivity?: '()' | '[)' | '(]' | '[]' | undefined) {
        return this._d.isBetween(from.getDto(), to.getDto(), granularity, inclusivity);
    }

    from(input: IkTime) {
        return this._d.from(input.getDto());
    }

    fromNow() {
        return this._d.fromNow();
    }

    local() {
        return new IkTime(this._d.tz(this.getTargetTimezone()), undefined, this._preset);
    }

    toString() {
        return this.getDto().toString();
    }

    getDto() {
        return this._d;
    }

    private getLocalizedFormat(format: string) {
        return this._preset.localizeFormat(format);
    }

    private getTargetTimezone() {
        return this.getLocaleSettings().target_tz;
    }

    private getLocaleSettings() {
        return this._preset.localeSettings();
    }
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function isTime(obj: any): obj is IkTime {
    return obj instanceof IkTime;
}
