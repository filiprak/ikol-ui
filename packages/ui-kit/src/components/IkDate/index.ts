import IkDate from './IkDate.vue';
export interface IkDate extends InstanceType<typeof IkDate> { };
export { IkDate };
