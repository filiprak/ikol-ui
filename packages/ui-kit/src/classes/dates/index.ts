import type { unitOfTime } from 'moment-timezone';
import type { IkTimeInput } from './IkTime';
import { IkTime, isTime } from './IkTime';
import { DATE_FORMATS } from './formats';
import { isObject } from '@ui/utils/helpers';
import { IkDuration } from './IkDuration';
import type { IkLocaleSettings } from '../locale/settings';
import { getDefaultLocaleSettings } from '../locale/settings';

export type IkTimeQuickOptions = 'today' | 'yesterday' | 'last_week' | 'last_month' | 'last_7_days' | 'this_week' | 'this_month';

function createLocalePreset(settings: Partial<IkLocaleSettings> = {}) {
    let preset: IkLocaleSettings;

    function time(input?: IkTimeInput, format?: string) {
        return new IkTime(input, format, time);
    }

    time.isTime = isTime;

    time.utc = (input?: IkTimeInput, format?: string) => {
        return IkTime.utc(input, format, time);
    };

    time.localeSettings = (): IkLocaleSettings => {
        if (!preset) {
            preset = Object.assign({}, getDefaultLocaleSettings(), settings);
        }
        return preset;
    };

    time.localizeFormat = (format: string) => {
        let l_format;
        const settings = time.localeSettings();

        if (format[0] === '!') {
            l_format = format.slice(2, format.length - 1);
        } else {
            if (DATE_FORMATS[settings.date_format] && DATE_FORMATS[settings.date_format][format]) {
                l_format = DATE_FORMATS[settings.date_format][format];
            } else {
                l_format = format;
            }
            // check if 12H/24H time format specification
            if (isObject(l_format)) {
                l_format = l_format[settings.time_format];
            }
        }
        return l_format;
    };

    time.isFullDaySpan = (start: IkTime, end: IkTime) => {
        if (start.isValid() && end.isValid()) {
            return (
                start.clone().startOf('day').isSame(start) &&
                (end.clone().startOf('day').isSame(end) || end.clone().endOf('day').add(1, 'second').isSame(end))
            );
        } else {
            return false;
        }
    };

    time.quickDatespans = (): Partial<Record<IkTimeQuickOptions, { label: string, value: { start: IkTime, end: IkTime } }>> => {
        return {
            today: {
                label: '{{_*en*Today_}}',
                value: { start: time().startOf('day'), end: time().endOf('day') },
            },
            yesterday: {
                label: '{{_*en*Yesterday_}}',
                value: { start: time().subtract(1, 'days').startOf('day'), end: time().subtract(1, 'days').endOf('day') },
            },
            last_week: {
                label: '{{_*en*Last week_}}',
                value: { start: time().subtract(1, 'weeks').startOf('week'), end: time().subtract(1, 'weeks').endOf('week') },
            },
            last_month: {
                label: '{{_*en*Last month_}}',
                value: { start: time().subtract(1, 'months').startOf('month'), end: time().subtract(1, 'months').endOf('month') },
            },
            last_7_days: {
                label: '{{_*en*Last 7 days_}}',
                value: { start: time().subtract(7, 'days').startOf('day'), end: time().endOf('day') },
            },
            this_week: {
                label: '{{_*en*This week_}}',
                value: { start: time().startOf('week'), end: time().endOf('day') },
            },
            this_month: {
                label: '{{_*en*This month_}}',
                value: { start: time().startOf('month'), end: time().endOf('day') },
            },
        };
    };

    time.quickDatespan = (name?: IkTimeQuickOptions) => {
        if (name) {
            const spans = time.quickDatespans()[name];
            if (spans) {
                return (spans.value || undefined);
            }
        }
        return undefined;
    };

    return time;
}

const time = createLocalePreset();

const duration = (input?: number | string, unit?: unitOfTime.DurationConstructor) => {
    return new IkDuration(input, unit);
};

export type IkTimePreset = typeof time;
export { time, duration, createLocalePreset };
