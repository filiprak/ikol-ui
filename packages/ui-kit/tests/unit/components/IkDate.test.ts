import MockDate from 'mockdate';
import type { RenderOptions } from '@ui/tests/tools/utils';
import { divStub, getBody, renderShallow, wait } from '@ui/tests/tools/utils';
import { IkDate } from '@ui/components/IkDate';
import { time } from '@ui/classes/dates';
import { IkTime } from '@ui/classes/dates/IkTime';
import { IkPopover } from '@ui/components/IkPopover';
import { IkInput } from '@ui/components/IkInput';

describe('IkDate', () => {
    const getInstance = (options: RenderOptions<typeof IkDate> = {}) => renderShallow(IkDate, {
        global: {
            renderStubDefaultSlot: true,
            stubs: { IkPopover, IkInput, IkScrollArea: divStub() },
        },
    }, options);

    beforeAll(() => {
        MockDate.set('2020-07-10 12:37:00');
    });

    afterAll(() => {
        MockDate.reset();
    });

    it('renders with default state', () => {
        const wrapper = getInstance({
            stubs: { IkPopover: true, IkInput: true, IkScrollArea: true },
        });
        expect(wrapper.classes()).toStrictEqual([
            'ik-date',
        ]);
        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders with value', () => {
        const wrapper = getInstance({
            props: {
                modelValue: time('2020-01-01'),
            },
        });

        expect(wrapper.findAllComponents(IkInput).at(0)?.props('modelValue')).toBe('Wed, 1 Jan 2020');
    });

    it('renders with value and time', () => {
        const wrapper = getInstance({
            props: {
                time: true,
                modelValue: time('2020-07-23 15:46'),
            },
        });

        expect(wrapper.findAllComponents(IkInput).at(0)?.props('modelValue')).toBe('Thu, 23 Jul 2020');
        expect(wrapper.findAllComponents(IkInput).at(1)?.props('modelValue')).toBe('15:46');
    });

    it('updates date on value change', async () => {
        const wrapper = getInstance({
            props: {
                time: true,
                modelValue: time('2020-07-23 15:46'),
            },
        });

        expect(wrapper.findAllComponents(IkInput).at(0)?.props('modelValue')).toBe('Thu, 23 Jul 2020');
        expect(wrapper.findAllComponents(IkInput).at(1)?.props('modelValue')).toBe('15:46');

        await wrapper.setProps({ modelValue: time('2019-03-12 13:52:00') });

        expect(wrapper.findAllComponents(IkInput).at(0)?.props('modelValue')).toBe('Tue, 12 Mar 2019');
        expect(wrapper.findAllComponents(IkInput).at(1)?.props('modelValue')).toBe('13:52');
    });

    it('updates date on value change 2', async () => {
        const wrapper = getInstance({
            props: {
                time: true,
                modelValue: time('2020-07-23 15:46'),
            },
        });

        expect(wrapper.findAllComponents(IkInput).at(0)?.props('modelValue')).toBe('Thu, 23 Jul 2020');
        expect(wrapper.findAllComponents(IkInput).at(1)?.props('modelValue')).toBe('15:46');

        await wrapper.setProps({ modelValue: time('2019-03-12 13:52:00') });

        expect(wrapper.findAllComponents(IkInput).at(0)?.props('modelValue')).toBe('Tue, 12 Mar 2019');
        expect(wrapper.findAllComponents(IkInput).at(1)?.props('modelValue')).toBe('13:52');

        expect(wrapper.emitted('update:modelValue')).toBeUndefined();
    });

    it('clears date on day input clear', () => {
        const wrapper = getInstance({
            props: {
                time: true,
                modelValue: time('2020-07-23 15:46'),
            },
        });

        wrapper.findAllComponents(IkInput).at(0)?.vm.$emit('update:modelValue', ' ');
        wrapper.findAllComponents(IkInput).at(0)?.vm.$emit('focusout');

        expect(wrapper.emitted('update:modelValue')).toStrictEqual([
            [null],
        ]);
    });

    it('updates date when typed by hand 1', () => {
        const wrapper = getInstance({
            props: {
                time: true,
                modelValue: time('2020-07-23 15:46'),
            },
        });

        wrapper.findAllComponents(IkInput).at(0)?.vm.$emit('update:modelValue', 'Tue, 12 Mar 2019');
        wrapper.findAllComponents(IkInput).at(0)?.vm.$emit('focusout');

        wrapper.findAllComponents(IkInput).at(0)?.vm.$emit('update:modelValue', 'Tue, 13 Mar 2019');
        wrapper.findAllComponents(IkInput).at(0)?.vm.$emit('focusout');

        wrapper.findAllComponents(IkInput).at(0)?.vm.$emit('update:modelValue', 'Tue, 13 Jan 2019');
        wrapper.findAllComponents(IkInput).at(0)?.vm.$emit('focusout');

        wrapper.findAllComponents(IkInput).at(0)?.vm.$emit('update:modelValue', 'Tue, 13 Jan 2020');
        wrapper.findAllComponents(IkInput).at(0)?.vm.$emit('focusout');

        expect(wrapper.emitted('update:modelValue')).toStrictEqual([
            [expect.any(IkTime)],
            [expect.any(IkTime)],
            [expect.any(IkTime)],
            [expect.any(IkTime)],
        ]);

        expect(wrapper.emitted('update:modelValue')?.[0][0]?.toString())
            .toStrictEqual('Tue Mar 12 2019 15:46:00 GMT+0100');

        expect(wrapper.emitted('update:modelValue')?.[1][0]?.toString())
            .toStrictEqual('Wed Mar 13 2019 15:46:00 GMT+0100');

        expect(wrapper.emitted('update:modelValue')?.[2][0]?.toString())
            .toStrictEqual('Sun Jan 13 2019 15:46:00 GMT+0100');

        expect(wrapper.emitted('update:modelValue')?.[3][0]?.toString())
            .toStrictEqual('Mon Jan 13 2020 15:46:00 GMT+0100');
    });

    it('updates date when typed by hand 2', () => {
        const wrapper = getInstance({
            props: {
                time: true,
                modelValue: time('2020-07-23 15:46'),
            },
        });

        wrapper.findAllComponents(IkInput).at(0)?.vm.$emit('update:modelValue', '310122');
        wrapper.findAllComponents(IkInput).at(0)?.vm.$emit('focusout');

        wrapper.findAllComponents(IkInput).at(0)?.vm.$emit('update:modelValue', '30012022');
        wrapper.findAllComponents(IkInput).at(0)?.vm.$emit('focusout');

        expect(wrapper.emitted('update:modelValue')).toStrictEqual([
            [expect.any(IkTime)],
            [expect.any(IkTime)],
        ]);

        expect(wrapper.emitted('update:modelValue')?.[0][0]?.toString())
            .toStrictEqual('Mon Jan 31 2022 15:46:00 GMT+0100');

        expect(wrapper.emitted('update:modelValue')?.[1][0]?.toString())
            .toStrictEqual('Sun Jan 30 2022 15:46:00 GMT+0100');
    });

    it('updates time when typed by hand', () => {
        const wrapper = getInstance({
            props: {
                time: true,
                modelValue: time('2020-07-23 15:46'),
            },
        });

        wrapper.findAllComponents(IkInput).at(1)?.vm.$emit('update:modelValue', '01:17');
        wrapper.findAllComponents(IkInput).at(1)?.vm.$emit('focusout');

        expect(wrapper.emitted('update:modelValue')).toStrictEqual([
            [expect.any(IkTime)],
        ]);
        expect(wrapper.emitted('update:modelValue')?.[0][0]?.toString())
            .toStrictEqual('Thu Jul 23 2020 01:17:00 GMT+0200');
    });

    it('does not update time when typed invalid', () => {
        const wrapper = getInstance({
            props: {
                time: true,
                modelValue: time('2020-07-23 15:46'),
            },
        });

        wrapper.findAllComponents(IkInput).at(1)?.vm.$emit('update:modelValue', '25:19');
        wrapper.findAllComponents(IkInput).at(1)?.vm.$emit('focusout');

        expect(wrapper.emitted('update:modelValue')).toBeUndefined();
    });

    it('does not update date when typed by hand invalid', () => {
        const wrapper = getInstance({
            props: {
                time: true,
                modelValue: time('2020-07-23 15:46'),
            },
        });

        wrapper.findAllComponents(IkInput).at(0)?.vm.$emit('update:modelValue', 'xxx');
        wrapper.findAllComponents(IkInput).at(0)?.vm.$emit('focusout');

        expect(wrapper.emitted('update:modelValue')).toBeUndefined();
    });

    it('updates date on day click', async () => {
        const wrapper = getInstance({
            props: {
                time: true,
                modelValue: time('2020-07-23 15:46'),
            },
        });

        await wrapper.vm.focusDay();
        await getBody().findAll('.ik-date__cell').at(17)?.trigger('click');

        expect(wrapper.emitted('update:modelValue')).toStrictEqual([
            [expect.any(IkTime)],
        ]);
        expect(wrapper.emitted('update:modelValue')?.[0][0]?.toString())
            .toStrictEqual('Mon Jul 06 2020 15:46:00 GMT+0200');
    });

    it('updates date on time pick', async () => {
        const wrapper = getInstance({
            props: {
                time: true,
                modelValue: time('2020-07-23 15:46'),
            },
        });

        await wrapper.vm.focusHour();
        await getBody().findAll('.ik-date__select-item').at(23)?.trigger('click');

        expect(wrapper.emitted('update:modelValue')).toStrictEqual([
            [expect.any(IkTime)],
        ]);
        expect(wrapper.emitted('update:modelValue')?.[0][0]?.toString())
            .toStrictEqual('Thu Jul 23 2020 11:30:00 GMT+0200');
    });

    it('allows navigate months', async () => {
        const wrapper = getInstance({
            props: {
                time: true,
                modelValue: time('2020-07-23 15:46'),
            },
        });

        await wrapper.vm.focusDay();
        await wrapper.vm.$nextTick();

        await getBody().findAll('.ik-date__cell > div').at(0)?.trigger('click');
        expect(getBody().find('.ik-date__nav').text()).toContain('June 2020');

        await getBody().findAll('.ik-date__cell > div').at(0)?.trigger('click');
        expect(getBody().find('.ik-date__nav').text()).toContain('May 2020');

        await getBody().findAll('.ik-date__cell > div').at(1)?.trigger('click');
        await getBody().findAll('.ik-date__cell > div').at(1)?.trigger('click');
        await getBody().findAll('.ik-date__cell > div').at(1)?.trigger('click');
        expect(getBody().find('.ik-date__nav').text()).toContain('August 2020');
    });

    it('highlights selected day on calendar', () => {
        const wrapper = getInstance({
            props: {
                time: true,
                modelValue: time('2020-07-23 15:46'),
            },
        });

        wrapper.vm.focusDay();

        expect(getBody().find('.ik-date__cell--active').text()).toContain('23');
    });

    it('highlights today on calendar', () => {
        const wrapper = getInstance({
            props: {
                time: true,
                modelValue: time('2020-07-23 15:46'),
            },
        });

        wrapper.vm.focusDay();

        expect(getBody().find('.ik-date__cell--today').text()).toContain('10');
    });

    it('valid timezone offset bound active date test', () => {
        const wrapper = getInstance({
            props: {
                time: true,
                modelValue: time.utc('2020-07-04 23:30'),
            },
        });

        wrapper.vm.focusDay();

        expect(getBody().find('.ik-date__cell--active').text()).toContain('5');
    });

    it('renders valid month days 1', () => {
        const wrapper = getInstance({
            props: {
                time: true,
                modelValue: time.utc('2020-09-30 22:00'),
            },
        });

        wrapper.vm.focusDay();

        expect(getBody().findAll('.ik-date__cell--weekday').at(0)?.text()).toContain('Sun');

        expect(getBody().find('.ik-date__cell--active').text()).toContain('1');
        expect(getBody().findAll('.ik-date__cell--offset')).toHaveLength(4);
        expect(getBody().findAll('.ik-date__cell--offset').at(0)?.text()).toContain('27');
        expect(getBody().findAll('.ik-date__cell--offset').at(3)?.text()).toContain('30');
    });

    it('renders valid month days 2', () => {
        const wrapper = getInstance({
            props: {
                time: true,
                modelValue: time('2020-12-31 23:59'),
            },
        });

        wrapper.vm.focusDay();

        expect(getBody().findAll('.ik-date__cell--weekday').at(0)?.text()).toContain('Sun');

        expect(getBody().find('.ik-date__cell--active').text()).toContain('31');
        expect(getBody().findAll('.ik-date__cell--offset')).toHaveLength(4);
        expect(getBody().findAll('.ik-date__cell--offset').at(0)?.text()).toContain('29');
        expect(getBody().findAll('.ik-date__cell--offset').at(3)?.text()).toContain('2');
    });

    it('works with autofocus', async () => {
        const wrapper = getInstance({
            props: {
                autofocus: true,
            },
            attachTo: document.body,
        });

        await wait(50);

        expect(wrapper.findComponent(IkInput).find('input').element).toBe(document.activeElement);
    });

    it('renders with hide date', () => {
        const wrapper = getInstance({
            props: {
                time_only: true,
            },
        });

        expect(wrapper.findAll('.ik-date__day-input').length).toBe(0);
    });

    it('renders with custom hour prompt interval', () => {
        const wrapper = getInstance({
            props: {
                prompt_hour_interval: 15,
                time: true,
            },
        });

        wrapper.vm.focusHour();

        expect(getBody().findAll('.ik-date__select-item').length).toBe(96);
        expect(getBody().findAll('.ik-date__select-item').at(0)?.text()).toBe('00:00');
        expect(getBody().findAll('.ik-date__select-item').at(1)?.text()).toBe('00:15');
        expect(getBody().findAll('.ik-date__select-item').at(6)?.text()).toBe('01:30');
    });

    it('renders with custom hour prompt interval 2', () => {
        const wrapper = getInstance({
            props: {
                prompt_hour_interval: 60,
                time: true,
            },
        });

        wrapper.vm.focusHour();

        expect(getBody().findAll('.ik-date__select-item').length).toBe(24);
        expect(getBody().findAll('.ik-date__select-item').at(0)?.text()).toBe('00:00');
        expect(getBody().findAll('.ik-date__select-item').at(1)?.text()).toBe('01:00');
        expect(getBody().findAll('.ik-date__select-item').at(6)?.text()).toBe('06:00');
    });

    it('renders full width', () => {
        const wrapper = getInstance({
            props: {
                full_width: true,
            },
        });

        expect(wrapper.classes()).toContain('ik-date--full-width');
    });

    it('prop disabled_datespand is formatted', async () => {
        const wrapper = getInstance({
            props: {
                disabled_datespans: [{ from: null, to: time('2020-12-31 23:59') }],
                time: true,
            },
            stubs: {
                IkPopover,
            },
        });
        await wrapper.vm.$nextTick();

        expect(wrapper.vm.normalized_disabled_datespans).toBeInstanceOf(Array);
        expect(wrapper.vm.normalized_disabled_datespans[0].from.toString()).toBe('Thu Jan 01 1970 01:00:00 GMT+0100');
        expect(wrapper.vm.normalized_disabled_datespans[0].to.toString()).toBe('Thu Dec 31 2020 23:59:00 GMT+0100');
    });

    it('props "disabled_datespans" assigns the css class in the appropriate cells', async () => {
        const wrapper = getInstance({
            props: {
                modelValue: time('2024-01-08'),
                disabled_datespans: [{ from: time('2024-01-01'), to: time('2024-01-03') }],
            },
        });
        await wrapper.vm.$nextTick();
        wrapper.vm.focusDay();

        expect(getBody().findAll('.ik-date__cell--disabled').length).toBe(3);

        expect(getBody().findAll('.ik-date__cell--disabled').at(0)?.text()).toBe('1');
        expect(getBody().findAll('.ik-date__cell--disabled').at(1)?.text()).toBe('2');
        expect(getBody().findAll('.ik-date__cell--disabled').at(2)?.text()).toBe('3');
    });

    it('props "disabled_datespans" allows you to disable several time spans', async () => {
        const wrapper = getInstance({
            props: {
                modelValue: time('2024-01-08'),
                disabled_datespans: [{ from: time('2024-01-01'), to: time('2024-01-03') }, { from: time('2024-01-05'), to: time('2024-01-07') }],
            },
        });
        await wrapper.vm.$nextTick();
        wrapper.vm.focusDay();

        expect(getBody().findAll('.ik-date__cell--disabled').length).toBe(6);

        expect(getBody().findAll('.ik-date__cell--disabled').at(0)?.text()).toBe('1');
        expect(getBody().findAll('.ik-date__cell--disabled').at(1)?.text()).toBe('2');
        expect(getBody().findAll('.ik-date__cell--disabled').at(2)?.text()).toBe('3');

        expect(getBody().findAll('.ik-date__cell--disabled').at(3)?.text()).toBe('5');
        expect(getBody().findAll('.ik-date__cell--disabled').at(4)?.text()).toBe('6');
        expect(getBody().findAll('.ik-date__cell--disabled').at(5)?.text()).toBe('7');
    });
});
