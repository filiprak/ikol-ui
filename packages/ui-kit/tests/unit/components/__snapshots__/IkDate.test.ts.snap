// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IkDate renders with default state 1`] = `
<div
  class="ik-date"
>
  <div
    class="ik-date__day-input"
  >
    <div
      class="ik-popover__root"
    >
      <div
        class="ik-input"
      >
        <!---->
        <input
          autocomplete="off"
          class="ik-input__native ik-f-input ik-f-input--size-sm ik-f-input--flat"
          placeholder="{{_*en*Choose date_}}"
          type="text"
          value=""
        />
        <!---->
      </div>
    </div>
  </div>
  <!--v-if-->
</div>
`;
